presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockBuilder.java

```diff
@@ -504,6 +504,12 @@ private static int getHashPosition(Block keyBlock, int position, MethodHandle ke
             throw new RuntimeException(throwable);
         }
 
-        return (int) Math.floorMod(hashCode, hashTableSize);
+
+        return getMod(hashCode, hashTableSize);
+    }
+
+    static int getMod(long hashcode, int hashTableSize)
+    {
+        return (int) (((Long.hashCode(hashcode) & 0x7fff_ffff) * (long) hashTableSize) >> 31);
```

Comment:
This is interesting. It's effectively computing a value between 0 and hashTableSize using fixed point math, by treating the hashcode as a number between 0 and 1 (in binary). So, effectively:

```
hashTableSize * (x / (2^31))
```

where `x` is a value between 0 and 2^31 (due to the `& 0x7fff_ffff`)

Please add an explanation of what this is doing, otherwise, it seems rather arbitrary.

Reply:
This is not as complicated as you mentioned. This is a standard compiler technique to avoid division and mod.